src/spec-configuration/containerFeaturesOCI.ts

```diff
@@ -0,0 +1,220 @@
+    output.write(`blob url: ${blobUrl}`, LogLevel.Trace);
+
+    const success = await getFeatureBlob(output, env, blobUrl, ociCacheDir, featCachePath, featureRef);
+
+    if (!success) {
+        output.write(`Failed to download package for ${featureSet.sourceInformation.featureRef.featureName}`, LogLevel.Error);
+        throw new Error(`Failed to download package for ${featureSet.sourceInformation.featureRef.featureName}`);
+    }
+
+    return true;
+}
+
+export async function getFeatureManifest(output: Log, env: NodeJS.ProcessEnv, url: string, featureRef: OCIFeatureRef): Promise<OCIManifest | undefined> {
+    try {
+        const headers = {
+            'user-agent': 'devcontainer',
+            'Authorization': await getAuthenticationToken(env, output, featureRef.registry, featureRef.id),
+            'Accept': 'application/vnd.oci.image.manifest.v1+json',
+        };
+
+        const options = {
+            type: 'GET',
+            url: url,
+            headers: headers
+        };
+
+        const response = await request(options, output);
+        const manifest: OCIManifest = JSON.parse(response.toString());
+
+        return manifest;
+    } catch (e) {
+        output.write(`error: ${e}`, LogLevel.Error);
+        return undefined;
+    }
+}
+
+// Downloads a blob from a registry.
+export async function getFeatureBlob(output: Log, env: NodeJS.ProcessEnv, url: string, ociCacheDir: string, featCachePath: string, featureRef: OCIFeatureRef): Promise<boolean> {
+    // TODO: Paralelize if multiple layers (not likely).
+    // TODO: Seeking might be needed if the size is too large.
+    try {
+        const tempTarballPath = path.join(ociCacheDir, 'blob.tar');
+
+        const headers = {
+            'user-agent': 'devcontainer',
+            'Authorization': await getAuthenticationToken(env, output, featureRef.registry, featureRef.id),
+            'Accept': 'application/vnd.oci.image.manifest.v1+json',
+        };
+
+        const options = {
+            type: 'GET',
+            url: url,
+            headers: headers
+        };
+
+        const blob = await request(options, output);
+
+        await mkdirpLocal(featCachePath);
+        await writeLocalFile(tempTarballPath, blob);
+        await tar.x(
+            {
+                file: tempTarballPath,
+                cwd: featCachePath,
+            }
+        );
+
+        return true;
+    } catch (e) {
+        output.write(`error: ${e}`, LogLevel.Error);
+        return false;
+    }
+}
+
+async function getAuthenticationToken(env: NodeJS.ProcessEnv, output: Log, registry: string, id: string): Promise<string> {
+    // TODO: Use operating system keychain to get credentials.
+    // TODO: Fallback to read docker config to get credentials.
+
+    // TEMP for ghcr.io
+    const githubToken = env['GITHUB_TOKEN'];
+    if (registry !== 'ghcr.io' || !githubToken) {
```

Comment:
Adding knowledge about `GITHUB_TOKEN` which makes this work for `ghcr.io` seamlessly in places like github actions is nice, but then we should consider adding support for env variables for other registries.

Reply:
Moving forward we'll want to read the OS keychain, etc to read the more standard .docker/config.json file.  

I do like this idea as temporary flexibility, or in cases where it would just be easier to set a secret (like in actions?)